crypto/src/key/hdkd/derivation_path.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key::hdkd::DerivationError;
+use crate::key::hdkd::DerivationError::*;
+use core::fmt;
+use std::fmt::Formatter;
+use std::slice::Iter;
+use std::str::FromStr;
+
+const PREFIX: &str = "m";
+const SEPARATOR: &str = "/";
+const HARDENED_APOS: char = '\'';
+const HARDENED_H: char = 'h';
+const HARDENED_BIT: u32 = 0x80000000;
+
+/// BIP32-like child numbers. Currently we only support hardened derivations
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub struct ChildNumber(DerivationType);
+
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+enum DerivationType {
+    Hardened(u32),
+}
+
+impl ChildNumber {
+    /// Return a hardened child number
+    pub fn hardened(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            Ok(ChildNumber(DerivationType::Hardened(index)))
+        } else {
+            Err(InvalidChildNumber(index))
+        }
+    }
+
+    /// Return a normal child number. Currently unsupported
+    fn normal(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            // For the time being we don't support non-hardened derivations
+            Err(UnsupportedDerivationType)
+        } else {
+            Err(InvalidChildNumber(index))
+        }
+    }
+
+    /// Return a BIP32-like child number index that has the hardened bit set if needed
+    pub fn to_encoded_index(&self) -> u32 {
+        match self.0 {
+            DerivationType::Hardened(i) => i | HARDENED_BIT,
+        }
+    }
+
+    /// Returns true if this child number is hardened
+    pub fn is_hardened(&self) -> bool {
+        match self.0 {
+            DerivationType::Hardened(_) => true,
+        }
+    }
+}
+
+impl FromStr for ChildNumber {
+    type Err = DerivationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !s.is_ascii() {
+            return Err(InvalidDerivationPathFormat);
+        }
+        // Check if this child number is hardened
+        if let Some(s) = s.strip_suffix([HARDENED_APOS, HARDENED_H]) {
+            ChildNumber::hardened(s.parse().map_err(|_| InvalidChildNumberFormat)?)
+        } else {
+            ChildNumber::normal(s.parse().map_err(|_| InvalidChildNumberFormat)?)
+        }
+    }
+}
+
+impl fmt::Display for ChildNumber {
```

Comment:
Do we want to instead implement ToString, and make this just call that?

Reply:
The way it's laid out in the standard library, it's the other way around. This is defined:
```rust
impl<T: fmt::Display + ?Sized> ToString for T
```

Meaning if you implement `fmt::Display`, you get `ToString` for free.

Reply:
Nice! Didn't know that! Cheers!